src/cantools/database/utils.py

```diff
@@ -123,15 +123,31 @@ def decode_data(data: bytes,
                 decode_choices: bool,
                 scaling: bool,
                 allow_truncated: bool,
+                allow_excess: bool,
                 ) -> SignalDictType:
 
     actual_length = len(data)
-    if allow_truncated and actual_length < expected_length:
-        data = data.ljust(expected_length, b"\xFF")
+
+    if actual_length < expected_length:
```

Comment:
Would you mind nesting both length checks to optimize for the happy path?

```python3
if len(data) != expected_length:
    if allow_truncated:
        data = data.ljust(expected_length, b"\xFF")
    if allow_excess:
        data = data[:expected_length]
    if len(data) != expected_length:
        raise ValueError
```

Reply:
done: 19207387d6a97779d9bff4ac9cd6a6d84489db66 (I've added an outer `if` to get more comprehensive error codes, though...)

Reply:
But the error message is the same for both cases? What's the difference? 🤔

Reply:
good point. changed: bb7055d7f4171d2aa57128b2761b2c4f67dd077d